### PR TITLE
add support for sourceip hashing in tcp proxy

### PR DESCRIPTION
**Pull request by ramaraochavali**
Fixes https://github.com/istio/istio/issues/33558

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [X] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by howardjohn** [comment]
We need to update the API docs as well. Also add an integration test to https://github.com/istio/istio/blob/f2909db45681292f4acaf550d83cc92f2be47828/tests/integration/pilot/common/routing.go#L1409.

The other thing is backwards compatibility. I think we are probably ok here, but it could represent a behavioral change for some users

File: pilot/pkg/networking/core/v1alpha3/networkfilter.go [hunk 1]
```diff
@@ -167,7 +174,9 @@ func buildOutboundNetworkFilters(node *model.Proxy,
 			statPrefix = util.BuildStatPrefix(push.Mesh.OutboundClusterStatName, routes[0].Destination.Host,
 				routes[0].Destination.Subset, port, &service.Attributes)
 		}
-		return buildOutboundNetworkFiltersWithSingleDestination(push, node, statPrefix, clusterName, port)
+		destRule := push.DestinationRule(node, service)
+		destinationRule := CastDestinationRule(destRule)
+		return buildOutboundNetworkFiltersWithSingleDestination(push, node, statPrefix, clusterName, port, destinationRule)
 	}
 	return buildOutboundNetworkFiltersWithWeightedClusters(node, routes, push, port, configMeta)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/networkfilter.go:181]: Doesn't it need src ip hash?

File: pilot/pkg/networking/core/v1alpha3/networkfilter.go [hunk 2]
```diff
@@ -69,18 +70,25 @@ func setAccessLogAndBuildTCPFilter(push *model.PushContext, config *tcp.TcpProxy
...
 	}
 
 	idleTimeout, err := time.ParseDuration(node.Metadata.IdleTimeout)
 	if err == nil {
 		tcpProxy.IdleTimeout = durationpb.New(idleTimeout)
 	}
 
+	// If destinationrule has consistent hash source ip set, use it for tcp proxy.
+	if destinationRule != nil && destinationRule.GetTrafficPolicy().GetLoadBalancer().GetConsistentHash() != nil &&
+		destinationRule.GetTrafficPolicy().GetLoadBalancer().GetConsistentHash().GetUseSourceIp() {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/networkfilter.go:86]: You should consider subset hash here

File: pilot/pkg/networking/core/v1alpha3/networkfilter.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/networkfilter.go:181]: I do not know if it makes sense for weighted clusters - Does it?

File: pilot/pkg/networking/core/v1alpha3/networkfilter.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/networkfilter.go:86]: Yes. I will add

**Comment by ramaraochavali** [comment]
> We need to update the API docs as well. Also add an integration test to

Yes. Planning to update API docs once this is merged. Will add an integration test

**Comment by ramaraochavali** [comment]
@hzxuzhonghu @howardjohn any more comments on this?

File: pilot/pkg/networking/core/v1alpha3/networkfilter.go [hunk 3]
```diff
@@ -69,18 +70,44 @@ func setAccessLogAndBuildTCPFilter(push *model.PushContext, config *tcp.TcpProxy
...
 		ClusterSpecifier: &tcp.TcpProxy_Cluster{Cluster: clusterName},
-		// TODO: Need to set other fields such as Idle timeouts
 	}
 
 	idleTimeout, err := time.ParseDuration(node.Metadata.IdleTimeout)
 	if err == nil {
 		tcpProxy.IdleTimeout = durationpb.New(idleTimeout)
 	}
 
+	if destinationRule != nil {
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/networkfilter.go:84]:
```suggestion
	if destinationRule.GetTrafficPolicy().GetLoadBalancer().GetConsistentHash().GetUseSourceIp() {
```

protobuf adds nil checks into each Get call

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/networkfilter.go:84]: nevermind it doesn't quite work due to subsets. But we can simplify L85-L86

File: pilot/pkg/networking/core/v1alpha3/networkfilter.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/networkfilter.go:200]: Don't we support it for HTTP? We should be consistent at least I think

**Review by howardjohn — COMMENTED** [review]
LGTM aside from weighted cluster concerns

File: pilot/pkg/networking/core/v1alpha3/networkfilter.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/networkfilter.go:200]: I just looked at envoy code and it picks the cluster based on weights first applies the hash - so we should be able to do it here as well. Let me try

**Review by howardjohn — APPROVED** [review]

**Pull request merged**